src/Internal/RestockRefundedItemsAdjuster.php

```diff
@@ -0,0 +1,67 @@
+<?php
+/**
+ * RestockRefundedItemsAdjuster class file.
+ */
+
+namespace Automattic\WooCommerce\Internal;
+
+use Automattic\WooCommerce\Proxies\LegacyProxy;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class to adjust or initialize the restock refunded items.
+ */
+class RestockRefundedItemsAdjuster {
+	/**
+	 * The order factory to use.
+	 *
+	 * @var WC_Order_Factory
+	 */
+	private $order_factory;
+
+	/**
+	 * Class initialization, to be executed when the class is resolved by the container.
+	 *
+	 * @internal
+	 */
+	final public function init() {
+		$this->order_factory = wc_get_container()->get( LegacyProxy::class )->get_instance_of( \WC_Order_Factory::class );
+		add_action( 'woocommerce_before_save_order_items', array( $this, 'initialize_restock_refunded_items' ), 10, 2 );
```

Comment:
So I have pushed up the changes you suggested. I've tested every possible scenario I can think of and so far it is working the way we want. As for your comment about the hook, I think it would be fine. Even if they called `wc_maybe_adjust_line_item_product_stock` directly it wouldn't effect anything as `_restock_refunded_items` initializes to zero if it isn't set. Because I think the intent is set that meta as soon as we can for older order versions.